### DISCUSSION TITLE
wire:loading not working when wire:click is added to button

### DISCUSSION THREAD

**Discussion by sukhcha-in** [livewire/livewire#3475] [7. Bugs (archived)]
### Description
`wire:loading` not working when `wire:click` is added to button
### Steps to reproduce
This is my button with two icons, one on loading state and cart icon on normal state:
```
<button class="w-full md:w-auto md:float-right bg-green-500 hover:bg-green-600 py-2 px-4 rounded-full text-white font-semibold text-lg focus:outline-none" wire:click="placeOrder" wire:loading.class="opacity-50 cursor-not-allowed" wire:loading.attr="disabled">
    <i class="icon-cart" wire:loading.remove></i> 
    <i class="icon-spin5 animate-spin" wire:loading></i> 
    Place Order
</button>
```
Cart icon gets removed and loading icon shows, but `wire:loading.class="opacity-50 cursor-not-allowed"` and `wire:loading.attr="disabled"` doesn't work at all.<br>
If i remove `wire:click="placeOrder"` from button it works fine.
**Context**
- Livewire version: v1.0.14, v1.1.0
- Laravel version: 7.0
- Browser: Chrome, Safari

**Comment by armancodes** [comment]
I tested this in Laravel 7 in both Chrome and Firefox using the v1.1.0 version and everything was working perfectly fine.
Try to do this in a fresh Laravel installation and check if it's a Livewire issue.

**Comment by flamisz** [comment]
I have the same issue. If I have 2 buttons, clicking on the button which has the `wire:click` work on both buttons, but clicking on the other one (which is a submit button) the `wire:loading` won't do anything on the button with `wire:click`.
The example:

```
<span class="inline-flex rounded-md shadow-sm">
    <button wire:click="cancel" wire:loading.class="bg-gray-200 cursor-default border-gray-200" wire:loading.class.remove="hover:border-yellow-200 border-gray-800" wire:loading.attr="disabled" type="button" class="w-full flex justify-center py-2 px-4 border-2 border-gray-800 font-bold uppercase tracking-wider hover:border-yellow-200 focus:outline-none focus:border-yellow-400 focus:shadow-outline-yellow transition duration-150 ease-in-out">
        Cancel
    </button>
</span>
<span class="ml-3 inline-flex rounded-md shadow-sm">
    <button wire:loading.class="bg-gray-200 cursor-default" wire:loading.class.remove="bg-yellow-200 hover:bg-gray-800 hover:text-yellow-200" wire:loading.attr="disabled" type="submit" class="w-full flex justify-center py-2 px-4 border border-transparent font-bold uppercase tracking-wider bg-yellow-200 hover:bg-gray-800 hover:text-yellow-200 focus:outline-none focus:border-yellow-400 focus:shadow-outline-yellow transition duration-150 ease-in-out">
        Save
    </button>
</span>
```

So clicking on the `Cancel` button works how it's expected, but clicking on the `Save` button only changes the `Save` button, but not the `Cancel` one.
I hope it's clear 😄 

I'll let you guys know if I figure out what's happening here.

**Comment by acurrieclark** [comment]
As a workaround to this problem, I have set the `wire:target` attribute to both of the relevant actions.

In @flamisz example, this would mean adding `wire:target="cancel, submit"` to the cancel button (assuming the `submit` button triggers an action called `submit`)

**Reply by akradim** [reply]
Hello and good day, I'm Mustafa . I had an important question about Livewire. I have a store and I wanted to share the links to my products on social networks for purchase. When people click on my links on social networks, the product should be added to their shopping cart without them being on my website.
now i need a link like esale/products?add-to-cart= just livewire wirk with ajax 
just i want example livewire link 
can you help me ? 
my email: akradim@hotmail.com

**Comment by flamisz** [comment]
> wire:target="cancel, submit"

That does the trick, thanks @acurrieclark . But I still believe it's a bug that should be solved.

**Comment by jeffochoa** [comment]
@calebporzio I had a different issue that might be linked to this.

Using two buttons inside a form:

```html
<form wire:submit.prevent="save">
    <button type="submit">send</button>

    <button type="button" wire:click="doSomething">click me</button>
</form>
```

The "click" on the second button works fine just the first time is clicked after the page loads, then stops working. I had to replace the button with a `<a href="#" wire:click >click me</a>` to make it work.

I hope this helps.

**Comment by cjmartino** [comment]
> I had to replace the button with a <a href="#" wire:click >click me</a> to make it work.

@calebporzio @jeffochoa Just wanted to note I came across the same issue. This works:

```html
<div class="col-md offset-md-2">
    <a href="#" class="btn btn-primary" wire:click.prevent="telSubmit" wire:loading.class="disabled">
        <span wire:loading wire:target="telSubmit" class="spinner-border spinner-border-sm" role="status" aria-hidden="true"></span>
        {{ __('Verify') }}
    </a>
    <button class="btn btn-outline-danger" wire:click="telCancel" wire:loading.attr="disabled">{{ __('Cancel') }}</button>
    <button class="btn btn-link" wire:click="telVerify('call')" wire:loading.attr="disabled">{{ __('Call me instead') }}</button>
</div>
```

This does not:

```html
<div class="col-md offset-md-2">
    <button class="btn btn-primary" wire:click="telSubmit" wire:loading.attr="disabled">
        <span wire:loading wire:target="telSubmit" class="spinner-border spinner-border-sm" role="status" aria-hidden="true"></span>
        {{ __('Verify') }}
    </button>
    <button class="btn btn-outline-danger" wire:click="telCancel" wire:loading.attr="disabled">{{ __('Cancel') }}</button>
    <button class="btn btn-link" wire:click="telVerify('call')" wire:loading.attr="disabled">{{ __('Call me instead') }}</button>
</div>
```

But with some further investigation, it seems like it was a problem with targeting the DOM. Adding some unique id attributes to the buttons fixed it:

```html
<div class="col-md offset-md-2">
    <button id="telSubmitter" href="#" class="btn btn-primary" wire:click.prevent="telSubmit" wire:loading.attr="disabled">
        <span wire:loading wire:target="telSubmit" class="spinner-border spinner-border-sm" role="status" aria-hidden="true"></span>
        {{ __('Verify') }}
    </button>
    <button id="telCanceller" class="btn btn-outline-danger" wire:click="telCancel" wire:loading.attr="disabled">{{ __('Cancel') }}</button>
    <button id="telVerifyTel" class="btn btn-link" wire:click="telVerify('call')" wire:loading.attr="disabled">{{ __('Call me instead') }}</button>
</div>
```

Hope that might get closer to the issue.

**Comment by squishythejellyfish** [comment]
👋 Oh Hi! I'm Squishy, the friendly jellyfish that manages Livewire issues.

I see this issue has been closed.

Here in the Livewire repo, we have an "issues can be closed guilt-free and without explanation" policy.

If for ANY reason you think this issue hasn't been resolved, PLEASE feel empowered to re-open it.

Re-opening actually helps us track which issues are a priority.

Reply "REOPEN" to this comment and we'll happily re-open it for you!

(More info on this philosophy here: https://twitter.com/calebporzio/status/1321864801295978497)

**Comment by booni3** [comment]
REOPEN

**Comment by booni3** [comment]
Adding wire:target and id's to each button element does fix this, however, I want to be able to disable by button during any load event, not just a targeted one. This bug is still present for this scenario.

My workaround for this currently, using the v2 syntax with Alpine JS is:

```blade
<button type="button"
        class="btn btn-primary"
        x-on:click="$wire.call('method')"
        wire:loading.attr="disabled"
>
    CLICK
</button>
```

**Comment by kallepyorala** [comment]
It seems that `wire:loading` works when the button itself is the initiator of the action but it is impossible to disable other buttons than the initiator. I did simple demonstration of this; https://github.com/azurinspire/livewire-loading (clone, composer install, php artisan serve). 

```html
<button wire:click="cancel" wire:loading.attr="disabled">Cancel</button>
<button wire:click="submit" wire:loading.attr="disabled">Submit</button>
```

When Cancel is clicked Submit is not disabled and other way around. Even adding `wire:loading.target="cancel, submit"` to the button doesn't help, only the button clicked is disabled.

**Comment by austenc** [comment]
@booni3 @azurinspire would one of you mind opening a new issue and including your Livewire + Alpine versions so we can better track the problem and get this fixed? Thanks!

**Comment by squishythejellyfish** [comment]
👋 Oh Hi! I'm Squishy, the friendly jellyfish that manages Livewire issues.

I see this issue has been closed.

Here in the Livewire repo, we have an "issues can be closed guilt-free and without explanation" policy.

If for ANY reason you think this issue hasn't been resolved, PLEASE feel empowered to re-open it.

Re-opening actually helps us track which issues are a priority.

Reply "REOPEN" to this comment and we'll happily re-open it for you!

(More info on this philosophy here: https://twitter.com/calebporzio/status/1321864801295978497)

**Comment by austenc** [comment]
Actually, looks like someone already did! Sorry to tag you both again, but could one of you confirm if this accurately represents your issue?

https://github.com/livewire/livewire/issues/2271

@booni3 
@azurinspire 

Thanks! 🙌

**Comment by suvires** [comment]
@calebporzio I have the same issue with a `<li>`:

`<li wire:loading.class="opacity-50" x-on:click="open = false" wire:click="$set('category_name', '{{ $category->name }}')" class="transition-colors cursor-pointer hover:bg-gray-100 pt-3 px-4 text-sm">`

If I delete wire:click event, wire:loading.class runs.

**Comment by xtfer** [comment]
For anyone else coming across this problem later, https://github.com/livewire/livewire/issues/1012#issuecomment-687573922 is a possible work around.

I used this to resolve an issue where I was replacing one button with another button, and the second button was becoming disabled.

**Comment by kondratovbr** [comment]
REOPEN

**Comment by kondratovbr** [comment]
> For anyone else coming across this problem later, [#1012 (comment)](https://github.com/livewire/livewire/issues/1012#issuecomment-687573922) is a possible work around.
> 
> I used this to resolve an issue where I was replacing one button with another button, and the second button was becoming disabled.

Tried adding IDs and `wire:key` on buttons - didn't work.

**Comment by joshhanley** [comment]
This issue has been converted to a discussion as we are now using discussions to track bugs and ideas.
See this announcement for more details https://github.com/livewire/livewire/discussions/3147

**Comment by rafalkrawiec** [comment]
Is there any progress on this issue? 

For now, once I specify for example `wire:click` directive to a button, I cannot simply use `wire:loading` on it to watch for actions other than defined in `wire:click`. Instead I have to use `wire:target` and pass all actions I have within my component to it.

This makes maintaining the code hard, sinceI need to make sure all my controls which I want to disable within the component have up-to-date actions list in `wire:target` directive.

For me there are two simple options there which could solve the issue:
1. Default behavior should be watching for all actions unless `wire:target` is declared. 
2. A modifier like `wire:loading.global` to force watching all actions on controls with for example `wire:click`.

First option makes most sense to me, but my approach is to disable all action controls within component when another action is running (to prevent user firing multiple actions). On the other hand a modifier doesn't change current behavior, so it's much safer as it won't affect existing projects.

**Reply by joshhanley** [reply]
@rafalkrawiec I ran into a similar issue myself with a button not too long ago, I think I just created a span inside the button and added the `wire:loading` to that instead of putting it on the button directly.

Hope this helps!

**Reply by alexanderchr** [reply]
Doesn’t help if you need to do something to the button, like disabling it
